pkg/document/document.go

```diff
@@ -1,60 +1,298 @@
+					Message:  fmt.Sprintf("symbol '%s' not found in %s", v[1], load.File),
+				})
+			}
+		}
+		for _, depload := range dep.Loads() {
+			for _, diag := range depload.Diagnostics {
+				diag.Range = load.Range
+				d.diagnostics = append(d.diagnostics, diag)
+			}
+		}
+	}
+}
+
+func (d *document) parseLoadStatements() {
+	nodes := query.LoadStatements(d.input, d.tree)
+	for _, n := range nodes {
+		parent := n.Parent()
+	parentloop:
+		for parent != nil {
+			switch parent.Type() {
+			case query.NodeTypeBlock, query.NodeTypeExpressionStatement:
+				parent = parent.Parent()
+			default:
+				break parentloop
+			}
+		}
+
+		if parent == d.tree.RootNode() {
+			load, diagnostics := loadStatement(d.input, n)
+			d.loads = append(d.loads, load)
+			d.diagnostics = append(d.diagnostics, diagnostics...)
+		} else {
+			d.diagnostics = append(d.diagnostics, protocol.Diagnostic{
+				Range:    query.NodeRange(n),
+				Severity: protocol.DiagnosticSeverityError,
+				Message:  fmt.Sprintf("load statement not allowed in %s", withArticle(strings.ReplaceAll(parent.Type(), "_", " "))),
+			})
+		}
+	}
+}
+
+func loadStatement(input []byte, n *sitter.Node) (LoadStatement, []protocol.Diagnostic) {
+	load := LoadStatement{Range: query.NodeRange(n)}
+	diagnostics := []protocol.Diagnostic{}
+	argsNode := n.ChildByFieldName("arguments")
+	notAString := func(nn *sitter.Node) protocol.Diagnostic {
+		return protocol.Diagnostic{
+			Range:    query.NodeRange(nn),
+			Severity: protocol.DiagnosticSeverityError,
+			Message:  fmt.Sprintf("load parameter must be a literal string, found '%s'", nn.Content(input)),
+		}
+	}
+	args := make([]*sitter.Node, argsNode.NamedChildCount())
+	for i := range args {
+		args[i] = argsNode.NamedChild(i)
+	}
+
+	if len(args) > 0 {
+		fileArg := args[0]
+		if fileArg.Type() == query.NodeTypeString {
+			load.File = unquote(fileArg.Content(input))
+		} else {
+			diagnostics = append(diagnostics, notAString(fileArg))
+		}
+	}
+
+	if len(args) > 1 {
+		for _, va := range args[1:] {
+			switch va.Type() {
+			case query.NodeTypeString:
+				s := unquote(va.Content(input))
+				load.Vars = append(load.Vars, [2]string{s, s})
+			case query.NodeTypeKeywordArgument:
+				alias := va.ChildByFieldName("name").Content(input)
+				nameNode := va.ChildByFieldName("value")
+				if nameNode.Type() == query.NodeTypeString {
+					load.Vars = append(load.Vars, [2]string{alias, unquote(nameNode.Content(input))})
+				} else {
+					diagnostics = append(diagnostics, notAString(nameNode))
+				}
+			default:
+				diagnostics = append(diagnostics, notAString(va))
+			}
+		}
+	} else {
+		diagnostics = append(diagnostics, protocol.Diagnostic{
+			Range:    query.NodeRange(n),
+			Severity: protocol.DiagnosticSeverityWarning,
+			Message:  "load statement did not specify any variables to import",
+		})
+	}
+	return load, diagnostics
+}
+
+func unquote(s string) string {
+	s, _ = strconv.Unquote(`"` + strings.Trim(s, s[0:1]) + `"`)
```

Comment:
Is there a way to get the actual, unquoted string from treesitter? Some of these starlark string literals seem like they'd fail here:
1. `r"foo"`
2. `"""foo"""`
3. `r"""foo"""` (apparently starlark-go accepts this, though I read the spec as not allowing it)
4. `'fo"o'`
5. `'fo\'o'`

(also maybe a good reason to handle the error)

Reply:
Ooh, yeah, those are good examples. I didn't find a way to get it from treesitter initially, but I'll have another look around and see what I can find out. At a minimum I can add tests for these, I admittedly didn't spend enough time thinking through edge cases for this.

Reply:
it occurred to me that getting docstrings is the same problem, so we could just reuse that solution, but it looks like it has similar problems:
https://github.com/tilt-dev/starlark-lsp/blob/c42183ababe2754eea0cf93c5050c527ce547235/pkg/query/symbol.go#L40